Add support for controlling light and fan motion, move sleep mode to be a switch

bdraco opened the pull request:
These are not mutually exclusive with the fan being
manually controlled like whoosh so it seems like a
switch better represents them.

HomeKit can now turn on / off the fan since it doesn't
think its already on because it would see the `PRESET_MODE_NONE`
as an `on` state.

bdraco committed:
Move sleep mode and auto to be a switch

These are not mutually exclusive with the fan being
manually controlled like whoosh

mikelawrence commented:
From what I can tell whoosh and sleep are mutually exclusive but the motion modes are not. I would think we could keep sleep as a preset. If we add the fan motion mode as a switch we should probably and the light motion mode too.

bdraco commented:
I’ll add the light motion switch later tonight or in the morning after sleep.  I don’t have a fan with a light so I can’t validate it though.

I just tested and if I set whoosh it overrides the speed.  If I set a speed it turns off whoosh.  I was able to set sleep mode and it didn’t turn off whoosh or change the speed and vise-versa.  I have the older models so maybe it behaves differently on the new ones?

bdraco commented:
So sleep mode is mutually exclusive of whoosh but not of a manual speed.  

Whoosh is mutually exclusive of a manual speed.

bdraco commented:
Even more confusing, turning on sleep mode disables whoosh but turning on whoosh does not disable sleep mode.

This is what I was missing in my testing.

bdraco commented:
## Modes Testing

| Mode | Mutually Exclusive of Speed (preset mode) | Mutually Exclusive of other Modes | Setting via API Disables other Modes |
| ---- | --------------------------- | -------------------------- | ---------------------------- |
| whoosh | True | True | False |
| sleep | False | True | True |
| motion (fan) | False | False | False |
| motion (light) | False | False | False |
| autocomfort | False | False | False |

bdraco commented:
... and yet even more confusing, turning on Whoosh when sleep mode is on disables auto comfort in the app

bdraco commented:
Looks like lights have sleep mode and motion as well.  Adjusted to handle that

Edit: Looks like lights only have motion from the picture of the remote I found.  Removed sleep mode from lights

bdraco commented:
I pulled the app up side by side and tested. Seems to match app behavior now.

bdraco commented:
HomeKit testing:

<img width="411" alt="Screen Shot 2021-03-03 at 8 33 01 PM" src="https://user-images.githubusercontent.com/663432/109921260-aceeca80-7c5f-11eb-9f25-b718e543d32f.png">

Looks good in HomeKit (switch is Whoosh) slider is speed.  Turning on Whoosh auto updates the slider, adjusting the slider will disable Whoosh as expected

mikelawrence commented:
I didn't like the first part of your table where you said that setting whoosh mode via API didn't disable other modes because in my experience that was not true. However when I took a closer look the problem is deeper. Sometimes when whoosh is set via the API the whoosh status is not reported by the fan; other times it works just fine. Not only that but if the fan is off when whoosh is set the fan should turn on automatically; this is also sporadic. I need to test further to see if its something I'm doing in the aiosenseme library.

I like where you are going with this PR but I need more time to test.

EDIT: Now I get what you are saying. If sleep is enabled setting whoosh doesn't do anything. However, setting whoosh when sleep is not enabled will turn on the fan. This is why you added the turn sleep mode off when turning on whoosh.

mikelawrence committed:
Add standalone light motion switch

Update documentation

mikelawrence commented:
It took me a while but I eventually understood the logic you presented. Your changes are solid. Thanks!

I added a motion control switch to the standalone light. I don't have one but the data I have indicates motion control exists on it too.

I am going to test a bit more before releasing.

mikelawrence committed:
Update validate.yml
